runtime/skiplist.c

```diff
@@ -0,0 +1,214 @@
+  } else {
+    return 0;
+  }
+}
+
+int caml_skiplist_find_below(struct skiplist * sk, uintnat k,
+                             uintnat * key, uintnat * data)
+{
+  int i;
+  struct skipcell * e, * f;
+
+  e = (struct skipcell *) sk;
+  for (i = sk->level; i >= 0; i--) {
+    while (1) {
+      f = e->forward[i];
+      if (f == NULL || f->key >= k + 1) break;
+      e = f;
+    }
+  }
+  if (e == (struct skipcell *) sk) {
+    return 0;
+  } else {
+    *key = e-> key; *data = e->data; return 1;
+  }
+}
+
+/* Insertion in a skip list */
+
+int caml_skiplist_insert(struct skiplist * sk,
+                         uintnat key, uintnat data)
+{
+  struct skipcell * update[NUM_LEVELS];
+  struct skipcell * e, * f;
+  int i, new_level;
+
+  CAMLassert(0 <= sk->level && sk->level < NUM_LEVELS);
+
+  /* Init "cursor" to list head */
+  e = (struct skipcell *) sk;
+  /* Find place to insert new node */
+  for (i = sk->level; i >= 0; i--) {
+    while (1) {
+      f = e->forward[i];
+      if (f == NULL || f->key >= key) break;
+      e = f;
+    }
+    update[i] = e;
+  }
+  e = e->forward[0];
+  /* If already present, update data */
+  if (e != NULL && e->key == key) {
+    e->data = data;
+    return 1;
+  }
+  /* Insert additional element, updating list level if necessary */
+  new_level = random_level();
+  if (new_level > sk->level) {
+    for (i = sk->level + 1; i <= new_level; i++)
+      update[i] = (struct skipcell *) sk;
+    sk->level = new_level;
+  }
+  e = caml_stat_alloc(SIZEOF_SKIPCELL +
+                      (new_level + 1) * sizeof(struct skipcell *));
+  e->key = key;
+  e->data = data;
+  for (i = 0; i <= new_level; i++) {
+    e->forward[i] = update[i]->forward[i];
+    update[i]->forward[i] = e;
+  }
+  return 0;
+}
+
+/* Deletion in a skip list */
+
+int caml_skiplist_remove(struct skiplist * sk, uintnat key)
+{
+  struct skipcell * update[NUM_LEVELS];
+  struct skipcell * e, * f;
+  int i;
+
+  CAMLassert(0 <= sk->level && sk->level < NUM_LEVELS);
+
+  /* Init "cursor" to list head */
+  e = (struct skipcell *) sk;
+  /* Find element in list */
+  for (i = sk->level; i >= 0; i--) {
+    while (1) {
+      f = e->forward[i];
+      if (f == NULL || f->key >= key) break;
+      e = f;
+    }
+    update[i] = e;
+  }
+  e = e->forward[0];
+  /* If not found, nothing to do */
+  if (e == NULL || e->key != key) return 0;
+  /* Rebuild list without node */
+  for (i = 0; i <= sk->level; i++) {
+    if (update[i]->forward[i] == e)
+      update[i]->forward[i] = e->forward[i];
+  }
+  /* Reclaim list element */
+  caml_stat_free(e);
+  /* Down-correct list level */
+  while (sk->level > 0 &&
+         sk->forward[sk->level] == NULL)
+    sk->level--;
+  return 1;
+}
+
+/* Iterate over a skip list */
+
+void caml_skiplist_iterate(void (*action)(uintnat, uintnat),
+                           struct skiplist * sk)
+{
+  struct skipcell * e;
+
+  for (e = sk->forward[0]; e != NULL; e = e->forward[0]) {
+    action(e->key, e->data);
+  }
```

Comment:
Couldn't we use `FOREACH_SKIPLIST_ELEMENT` here?

Reply:
No longer relevant.